### DISCUSSION TITLE
Easier CLI integration with CI processes

### DISCUSSION THREAD

**Discussion by slashben** [kubescape/kubescape#550] [Ideas]
Got some feedback from people about using Kubescape in CI.

They want to use it to block PR in case the YAMLs/HELM chart are not in line with a security baseline. Aiming the security threshold at 0 (no risk) and trying to use it as a security baseline is very hard:

The user needs to:
* run Kubescape with threshold 0
* create an exceptions YAML
* define an exception for each issue
* or fix it

The last two step can repeat for very long depending on the number of resources and controls in the framework.

I suggest a way in CLI to be able to focus on the important issues only. I suggest Kubescape CLI return errors if there are more failed resources than the number specified.

`--threshold-critical-resources <max number of critical resources with issues>`
`--threshold-high-resources <max number of high resources with issues>`
`--threshold-medium-resources <max number of medium resources with issues>`
`--threshold-low-resources <max number of low resources with issues>`

Of course, these has to take into account the excpetions.

Comments are welcome :)

**Comment by dwertent** [comment]
I agree with the problem, in the current state it is hard to actually define the correct threshold. The best way of setting a threshold in the current version is by pre scanning and pre fixing the Kubernetes manifests and setting the threshold based on the results.

What you are proposing will give the user a more flexible solution but it does not solve the core of the problem.
If we move on with your proposal I suggest as follows:
1. The different severities will act as different levels (similar to log levels) this way, when setting `--threshold-medium-resources=2` and the results include 3 high severities the scan will fail
2. We should not use numbers, we should use risk scan or any other calculation. This was a decision we made in the past so replicas will be calculated in a different way than multiple workloads

This being said, I think the solution is not by setting different thresholds but by assisting the users. We can assist the users by generating fixed YAMLs and auto generating exceptions files.

**Reply by slashben** [reply]
_"The different severities will act as different levels (similar to log levels) this way, when setting --threshold-medium-resources=2 and the results include 3 high severities the scan will fail"_
This is an interesting and somewhat understandable proposal. However, I don't like the implicit nature of it. Your proposal would cause that these arguments cannot coexist. If I said I agree to accept 3 high issues and 10 medium, what would be the equation? 
Rather make it mechanism explicit, even if the user needs to add more than one argument.

**Reply by slashben** [reply]
> 2. We should not use numbers, we should use risk scan or any other calculation. This was a decision we made in the past so replicas will be calculated in a different way than multiple workloads

We have been down in this road, especially when scanning in CI (YAML file and HELM) I don't see a problem here counting resources.

**Reply by slashben** [reply]
> This being said, I think the solution is not by setting different thresholds but by assisting the users. We can assist the users by generating fixed YAMLs and auto generating exceptions files.

Good idea, I don't see contradiction :)
